crates/forge/tests/e2e/trace_resources.rs

```diff
@@ -0,0 +1,325 @@
+}
+
+fn deserialize_call_trace(test_name: &str, temp_dir: &TempDir) -> ProfilerCallTrace {
+    let trace_data = fs::read_to_string(
+        temp_dir
+            .join(TRACE_DIR)
+            .join(format!("tests::{test_name}::{test_name}.json")),
+    )
+    .unwrap();
+    serde_json::from_str(&trace_data).expect("Failed to parse call trace")
+}
+
+fn ensure_resources_are_correct(call_trace: &ProfilerCallTrace) -> &ProfilerExecutionResources {
+    let mut child_resources = vec![];
+    for call in &call_trace.nested_calls {
+        child_resources.push(ensure_resources_are_correct(call));
+    }
+
+    let mut sum_child_resources = ProfilerExecutionResources::default();
+    for resource in child_resources {
+        sum_child_resources += resource;
+    }
+
+    let current_resources = &call_trace.used_resources.execution_resources;
+    assert!(current_resources.gt_eq_than(&sum_child_resources));
+    let resource_diff = current_resources - &sum_child_resources;
+    assert_correct_diff_for_easily_countable_syscalls(&resource_diff);
+
+    &call_trace.used_resources.execution_resources
+}
+
+fn assert_correct_diff_for_easily_countable_syscalls(resource_diff: &ProfilerExecutionResources) {
+    for syscall in [
+        EmitEvent,
+        GetBlockHash,
+        GetExecutionInfo,
+        StorageWrite,
+        StorageRead,
+        SendMessageToL1,
+        Keccak,
+    ] {
+        assert_eq!(
+            *resource_diff
+                .syscall_counter
+                .get(&syscall)
+                .unwrap_or_else(|| panic!("Expected resource diff to contain {syscall:?}")),
+            1,
+            "Incorrect diff for {syscall:?}"
+        );
+    }
+
+    for builtin in [
+        "poseidon_builtin",
+        "ec_op_builtin",
+        "bitwise_builtin",
+        "pedersen_builtin",
+    ] {
+        assert_eq!(
+            *resource_diff
+                .vm_resources
+                .builtin_instance_counter
+                .get(builtin)
+                .unwrap_or_else(|| panic!("Expected resource diff to contain {builtin:?}")),
+            1,
+            "Incorrect diff for {builtin:?}"
+        );
+    }
+}
+
+// When sth fails in the functions below and you didn't change anything in the cairo code it is a BUG.
+// If you changed the corresponding cairo code count the expected occurrences of syscalls manually first, then assert them.
+// TL;DR: DON't mindlessly change numbers to fix the tests if they ever fail.
+fn ensure_test_deploy_syscalls(temp_dir: &TempDir) {
+    let test_call_trace = deserialize_call_trace("test_call", temp_dir);
+    let top_call_syscalls = test_call_trace
+        .used_resources
+        .execution_resources
+        .syscall_counter;
+    assert_eq!(*top_call_syscalls.get(&Deploy).unwrap(), 11); // FIXME in #1631 (should be 14)
+    assert_eq!(*top_call_syscalls.get(&CallContract).unwrap(), 4); // FIXME in #1631 (should be 8)
+    assert_eq!(*top_call_syscalls.get(&LibraryCall).unwrap(), 1);
```

Comment:
I think we could try extracting some helper function here, because this code is repeated few times

Reply:
Definitely can create some macros / helper functions

Reply:
Will do